### ISSUE TITLE
AlphaVantage raises "Please input a valid date range" for previously working code

### ISSUE THREAD

**Issue by schneiderfelipe** [pydata/pandas-datareader#662]
Hi all,

The following snippet was working fine until I updated pandas-datareader:

```python3
start = datetime.datetime(1990, 5, 16)
end = datetime.datetime.now()
data = web.DataReader("ITUB3.SAO", "av-daily-adjusted", start=start, end=end)
```

Now it throws a `ValueError` ("Please input a valid date range"). Even leaving `start` and `end` as `None` doesn't work.

I believe it has to do with the indexing at line 81 of `pandas_datareader.av.time_series`:

https://github.com/pydata/pandas-datareader/blob/8eb5655ff90ed043f9297b9585c8ca2528f5c84c/pandas_datareader/av/time_series.py#L81

What do you think? Thanks in advance.

**Comment by dahifi** [comment]
Looks like it's still an issue @addisonlynch. Using the following code from the documentation: 
``f = web.DataReader("AAPL", "av-daily", 
                   start=datetime(2017, 2, 9), 
                   end=datetime(2017, 5, 24), 
                   api_key=os.getenv('ALPHAVANTAGE_API_KEY')
                  )

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-fd942bc63d60> in <module>
      2                    start=datetime(2017, 2, 9),
      3                    end=datetime(2017, 5, 24),
----> 4                    api_key=os.getenv('ALPHAVANTAGE_API_KEY')
      5                   )

~/.local/share/virtualenvs/tradePlan-24nzYghD/lib/python3.5/site-packages/pandas/util/_decorators.py in wrapper(*args, **kwargs)
    186                 else:
    187                     kwargs[new_arg_name] = new_arg_value
--> 188             return func(*args, **kwargs)
    189         return wrapper
    190     return _deprecate_kwarg

~/.local/share/virtualenvs/tradePlan-24nzYghD/lib/python3.5/site-packages/pandas_datareader/data.py in DataReader(name, data_source, start, end, retry_count, pause, session, api_key)
    577             pause=pause,
    578             session=session,
--> 579             api_key=api_key,
    580         ).read()
    581 

~/.local/share/virtualenvs/tradePlan-24nzYghD/lib/python3.5/site-packages/pandas_datareader/base.py in read(self)
     98         """Read data from connector"""
     99         try:
--> 100             return self._read_one_data(self.url, self.params)
    101         finally:
    102             self.close()

~/.local/share/virtualenvs/tradePlan-24nzYghD/lib/python3.5/site-packages/pandas_datareader/base.py in _read_one_data(self, url, params)
    110         else:
    111             raise NotImplementedError(self._format)
--> 112         return self._read_lines(out)
    113 
    114     def _read_url_as_StringIO(self, url, params=None):

~/.local/share/virtualenvs/tradePlan-24nzYghD/lib/python3.5/site-packages/pandas_datareader/av/time_series.py in _read_lines(self, out)
    110         data = data.loc[start_str:end_str]
    111         if data.empty:
--> 112             raise ValueError("Please input a valid date range")
    113         else:
    114             for column in data.columns:

ValueError: Please input a valid date range
```

Interestingly enough, I only noticed this error when I moved my Google Colab code to a local machine. So the above code still works on 0.7.4, but not on 0.8.1.

**Comment by austinteague** [comment]
I am still receiving the same error as noted by others above.

**Comment by dahifi** [comment]
I did email AlphaVantage over two weeks ago. I will followup with them and see if there's been a response.

**Comment by peh-empire** [comment]
I am experiencing a similar behavior when accessing from inside a docker container

I.e

`sudo docker run -e API_KEY=${API_KEY} -v output:/out -t <img-with-cmd>`

gives

```
INFO:root:Getting historical data for 7203.T
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.alphavantage.co:443
DEBUG:urllib3.connectionpool:https://www.alphavantage.co:443 "GET /query?function=TIME_SERIES_WEEKLY_ADJUSTED&symbol=7203.T&apikey=VDTULX0I2L5WUEI4&outputsize=full HTTP/1.1" 200 None

INFO:root:Could not read symbol 7203.T from None to None (Please input a valid date range)
INFO:root:sleep 10
INFO:root:Getting historical data for AAPL
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.alphavantage.co:443
DEBUG:urllib3.connectionpool:https://www.alphavantage.co:443 "GET /query?function=TIME_SERIES_WEEKLY_ADJUSTED&symbol=AAPL&apikey=VDTULX0I2L5WUEI4&outputsize=full HTTP/1.1" 200 None
INFO:root:Could not read symbol AAPL from None to None (Please input a valid date range)
```

when the same `cmd` is executed locally:

```
INFO:root:Getting historical data for 7203.T
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.alphavantage.co:443
DEBUG:urllib3.connectionpool:https://www.alphavantage.co:443 "GET /query?symbol=7203.T&function=TIME_SERIES_WEEKLY_ADJUSTED&apikey=VDTULX0I2L5WUEI4&outputsize=full HTTP/1.1" 200 None
INFO:root:sleep 10
INFO:root:Getting historical data for AAPL
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.alphavantage.co:443
DEBUG:urllib3.connectionpool:https://www.alphavantage.co:443 "GET /query?symbol=AAPL&function=TIME_SERIES_WEEKLY_ADJUSTED&apikey=VDTULX0I2L5WUEI4&outputsize=full HTTP/1.1" 200 None
```

The only difference is that the query is twisted but i verified that this does not cause any problem

**Comment by tobiaslocker** [comment]
I am experiencing a similar behavior when accessing from inside a docker container

I.e

`sudo docker run -e API_KEY=${API_KEY} -v output:/out -t <img-with-cmd>`

gives

```
INFO:root:Getting historical data for 7203.T
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.alphavantage.co:443
DEBUG:urllib3.connectionpool:https://www.alphavantage.co:443 "GET /query?function=TIME_SERIES_WEEKLY_ADJUSTED&symbol=7203.T&apikey=*********************&outputsize=full HTTP/1.1" 200 None

INFO:root:Could not read symbol 7203.T from None to None (Please input a valid date range)
INFO:root:sleep 10
INFO:root:Getting historical data for AAPL
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.alphavantage.co:443
DEBUG:urllib3.connectionpool:https://www.alphavantage.co:443 "GET /query?function=TIME_SERIES_WEEKLY_ADJUSTED&symbol=AAPL&apikey=*********************&outputsize=full HTTP/1.1" 200 None

**Comment by lsiqueirap** [comment]
I'm still receiving the error

**Comment by temirov** [comment]
Not sure if it's a good report but just in case:

```bash
sh-4.2$ pip show pandas_datareader
Name: pandas-datareader
Version: 0.9.0.dev0+15.g63a51b4
Summary: Data readers extracted from the pandas codebase,should be compatible with recent pandas versions
Home-page: https://github.com/pydata/pandas-datareader
Author: The PyData Development Team
Author-email: pydata@googlegroups.com
License: BSD License
Location: /home/ec2-user/.local/lib/python3.6/site-packages
Requires: requests, lxml, pandas
```

```python
av = pdr.DataReader(
    symbol, 
    "av-daily-adjusted", 
    '2019-10-11',
    '2019-10-12',
    api_key=AV_API_KEY
)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-48d358e5e414> in <module>()
     13     '2019-10-11',
     14     '2019-10-12',
---> 15     api_key=AV_API_KEY
     16 )
     17 av.describe()

~/anaconda3/envs/python3/lib/python3.6/site-packages/pandas/util/_decorators.py in wrapper(*args, **kwargs)
    186                 else:
    187                     kwargs[new_arg_name] = new_arg_value
--> 188             return func(*args, **kwargs)
    189         return wrapper
    190     return _deprecate_kwarg

~/.local/lib/python3.6/site-packages/pandas_datareader/data.py in DataReader(name, data_source, start, end, retry_count, pause, session, api_key)
    589             pause=pause,
    590             session=session,
--> 591             api_key=api_key,
    592         ).read()
    593 

~/.local/lib/python3.6/site-packages/pandas_datareader/base.py in read(self)
     98         """Read data from connector"""
     99         try:
--> 100             return self._read_one_data(self.url, self.params)
    101         finally:
    102             self.close()

~/.local/lib/python3.6/site-packages/pandas_datareader/base.py in _read_one_data(self, url, params)
    110         else:
    111             raise NotImplementedError(self._format)
--> 112         return self._read_lines(out)
    113 
    114     def _read_url_as_StringIO(self, url, params=None):

~/.local/lib/python3.6/site-packages/pandas_datareader/av/time_series.py in _read_lines(self, out)
    110         data = data.loc[start_str:end_str]
    111         if data.empty:
--> 112             raise ValueError("Please input a valid date range")
    113         else:
    114             for column in data.columns:

ValueError: Please input a valid date range
```

**Comment by temirov** [comment]
Let's reopen this issue @dahifi @bashtage , seems like it hasn't been fixed

**Comment by schneiderfelipe** [comment]
> Let's reopen this issue @dahifi @bashtage , seems like it hasn't been fixed

@temirov I agree this should be reopened due to so many people facing the same issue, but I must point out that I'm currently using version 0.7.4 with Python 3.6 and I face no issues.

At the end of `pandas_datareader/av/time_series.py` I have the following:

```python3
...
class AVTimeSeriesReader(AlphaVantage):
...
    def _read_lines(self, out):
        data = super(AVTimeSeriesReader, self)._read_lines(out)
        data = data[::-1]
        start_str = self.start.strftime('%Y-%m-%d')
        end_str = self.end.strftime('%Y-%m-%d')
        data = data.loc[start_str:end_str]
        if data.empty:
            raise ValueError("Please input a valid date range")
        else:
            for column in data.columns:
                if column == 'volume':
                    data[column] = data[column].astype('int64')
                else:
                    data[column] = data[column].astype('float64')
        return data
```

So the entries is properly inverted at the second line of `_read_lines`. That's pretty much what #666 [has changed](https://github.com/addisonlynch/pandas-datareader/blame/283a41d7bcec659b060fc811e906e2b341d948eb/pandas_datareader/av/time_series.py#L80).

**Comment by temirov** [comment]
@schneiderfelipe Thank you for sharing the code!

That is interesting that it works for you, because it doesn't work for me and the code looks the same

```bash
sh-4.2$ pip show pandas_datareaderName: pandas-datareader
Version: 0.8.1
Summary: Data readers extracted from the pandas codebase,should be compatible with recent pandas versions
Home-page: https://github.com/pydata/pandas-datareader
Author: The PyData Development Team
Author-email: pydata@googlegroups.com
License: BSD License
Location: /home/ec2-user/.local/lib/python3.6/site-packages

sh-4.2$ tail -n20 /home/ec2-user/.local/lib/python3.6/site-packages/pandas_datareader/av/time_series.py
        if self.intraday:
            p.update({"interval": "1min"})
        return p

    def _read_lines(self, out):
        data = super(AVTimeSeriesReader, self)._read_lines(out)
        # reverse since alphavantage returns descending by date
        data = data[::-1]
        start_str = self.start.strftime("%Y-%m-%d")
        end_str = self.end.strftime("%Y-%m-%d")
        data = data.loc[start_str:end_str]
        if data.empty:
            raise ValueError("Please input a valid date range")
        else:
            for column in data.columns:
                if column == "volume":
                    data[column] = data[column].astype("int64")
                else:
                    data[column] = data[column].astype("float64")
        return data
```

**Comment by schneiderfelipe** [comment]
> That is interesting that it works for you, because it doesn't work for me and the code looks the same

> I can confirm that commenting out the inversion of the range makes code work:

Interesting. Is the data being reversed somewhere else?

**Comment by addisonlynch** [comment]
It appears that the data format is changing on the provider's end. It would be helpful to know what the long-term plan is on this from AV because it seems to have reversed a couple times.

**Comment by temirov** [comment]
@addisonlynch I can add a condition to see which direction the data is going and reverse it if needed ;-) Just need to brush up my python testing skills ...

**Comment by addisonlynch** [comment]
Thanks @temirov that would be helpful, however I think it may be a good idea to first determine the root cause of this. I haven't been able to look into it yet but I would be surprised if the provider were arbitrarily reversing its time series data (one of its most popular endpoints). There may be another issue at hand (and maybe not 😄)

**Comment by schneiderfelipe** [comment]
> Thanks @temirov that would be helpful, however I think it may be a good idea to first determine the root cause of this. I haven't been able to look into it yet but I would be surprised if the provider were arbitrarily reversing its time series data (one of its most popular endpoints). There may be another issue at hand (and maybe not smile)

A possibility would be that their data is simply not standardized. Some tickers may come out reversed, while others don't. It seems that the ticker (`ITUB3.SAO`) that made me open this issue is working fine with the current code.

**Comment by NRVA** [comment]
I belive I find the same problem with data from Fred:

```
from datetime import datetime
start = datetime.datetime(2010,1,1)
web.DataReader("BAMLH0A0HYM2", "fred", start)
```

returning:

```
Exception in thread Thread-612:
Traceback (most recent call last):
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/site-packages/multitasking/__init__.py", line 102, in _run_via_pool
    return callee(*args, **kwargs)
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/site-packages/yfinance/multi.py", line 167, in _download_one_threaded
    actions, period, interval, prepost, proxy, rounding)
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/site-packages/yfinance/multi.py", line 182, in _download_one
    rounding=rounding, many=True)
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/site-packages/yfinance/base.py", line 115, in history
    _time.strptime(str(start), '%Y-%m-%d')))
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/_strptime.py", line 571, in _strptime_time
    tt = _strptime(data_string, format)[0]
  File "/Users/familien_nerva-koch/opt/anaconda3/lib/python3.7/_strptime.py", line 359, in _strptime
    (data_string, format))
ValueError: time data 'fred' does not match format '%Y-%m-%d'
```

**Comment by arunkhattri** [comment]
i am still getting an ValueError: Please input a valid date range
pandas_datareader version: 0.8.1
python version: 3.7.4

trying to run example snippet

`aapl = DataReader("AAPL", "av-daily", start=datetime(2017, 2, 9), end=datetime(2017, 5, 24), api_key=config['Alpha_Vantage']['key'])`

`aapl.loc["2017-02-09"]`

**Comment by hmonteiro** [comment]
Is there even a reason why it is assumed the data is sorted in the first place? If the data comes with the proper dates attached, why do we even care? Couldn't we just sort it before we slice it?

**Comment by scyap** [comment]
```
  File "C:\Users\your_user_name\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pandas_datareader\av\time_series.py", line 114, in _read_lines
    raise ValueError("Please input a valid date range")
ValueError: Please input a valid date range
```

This is the error message I see. So I drill into the library file and see what when wrong.
Follow the file path in the error message and open up the file:
......\lib\site-packages\pandas_datareader\av\**time_series.py**

**Problem:**
`data.loc[start_str:end_str]`
This line of code will return Empty DataFrame 

**Solution:**
Sort the index first before using the .loc[]
Reference: [Pandas DataFrame.loc returns Empty DataFrame](https://stackoverflow.com/questions/52287899/pandas-dataframe-loc-returns-empty-dataframe/52288027)

**Action:**
add this line of code 
`data = data.sort_index()`
into the original as below

```
    def _read_lines(self, out):
        data = super(AVTimeSeriesReader, self)._read_lines(out)
        # reverse since alphavantage returns descending by date
        data = data[::-1]
        **data = data.sort_index() # sort index first, so that data.loc[] works**
        start_str = self.start.strftime("%Y-%m-%d")
        end_str = self.end.strftime("%Y-%m-%d")
        data = data.loc[start_str:end_str]
        if data.empty:
            raise ValueError("Please input a valid date range")
        else:
            for column in data.columns:
                if column == "volume":
                    data[column] = data[column].astype("int64")
                else:
                    data[column] = data[column].astype("float64")
        return data
```
